### PR TITLE
[#7918] replica_truncate: Fix JSON output for some cases (4-3-stable)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #7918

Unit tests are passing. Also confirmed that the replica_truncate unit tests fail as expected before the change. That is, successful truncates return `nullptr` for the output string before the changes, and after the changes they correctly return a message and replica information.

While the writing tests, I noticed one case which produced results which differed from my expectations and I went ahead and changed this as well in the API. I feel like not everybody may agree with the decision, so I wanted to highlight the test case section named `resource_name_with_no_replica` for review/discussion.

Running core tests now.

### CODE REVIEW

File: unit_tests/src/test_replica_truncate.cpp [hunk 1]
```diff
@@ -716,47 +970,158 @@ TEST_CASE("inputs_that_will_not_work")
...
+        // to truncate does not exist. Even though an existing replica may have won the vote, it was not the intended
+        // target of the truncation and has nothing to do with the failure. Therefore, we expect no replica information.
+        const auto error_message = fmt::format("rs_replica_truncate: Hierarchy descending from specified resource name "
+                                               "[{}] does not have a replica of [{}] "
+                                               "or the replica is inaccessible at this time.",
+                                               test_resc.c_str(),
+                                               target_object.c_str());
+        const auto expected_json_output = nlohmann::json{{"message", error_message},
+                                                         {"replica_number", nlohmann::json::value_t::null},
+                                                         {"resource_hierarchy", nlohmann::json::value_t::null}};
```
Comment (alanking) [inline unit_tests/src/test_replica_truncate.cpp:1075]: This is the assertion that I wanted to draw attention to.

**Comment by alanking** [comment]
Force-push due to wrong issue number. Branch name will remain incorrect. :( Sorry.

**Review by korydraughn — COMMENTED** [review]
Overall, the changes look good.

Will now look at the assertion you highlighted.

File: unit_tests/src/test_replica_truncate.cpp [hunk 2]
```diff
@@ -122,6 +126,18 @@ TEST_CASE("basic_two_standalone_resources")
         // Attempt to truncate the object.
         REQUIRE(0 == rc_replica_truncate(&comm, &truncate_doi, &output_str));
 
+        // Ensure that the returned output structure has the expected contents.
+        REQUIRE(nullptr != output_str);
+        nlohmann::json json_out;
+        REQUIRE_NOTHROW(json_out = nlohmann::json::parse(output_str));
```
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:132]: The API never returns a string that is not a `nlohmann::json::dump`, so I'm not sure what you mean. Can you elaborate?

File: unit_tests/src/test_replica_truncate.cpp (hunk 1 above)
Comment (korydraughn) [reply unit_tests/src/test_replica_truncate.cpp:1075]: The new behavior is correct to me.

User requests a replica which doesn't exist or isn't under the target resource, therefore they get an error and no replica information.

File: unit_tests/src/test_replica_truncate.cpp [hunk 3]
```diff
@@ -319,6 +387,39 @@ TEST_CASE("two_replicas_in_replication_resource")
...
+        // determined that it was the same size as the requested truncate size. Check to make sure the replica info
+        // matches what we expect.
+
+        // We cannot expect one replica to have been truncated any more than the other. We must figure out which replica
+        // was truncated here, by number, and then we can assert which resource hierarchy we expect to see.
+        const auto replica_number_itr = json_out.find("replica_number");
+        REQUIRE(json_out.end() != replica_number_itr);
+        const auto replica_number = replica_number_itr->get<int>();
+        // This is a separate variable because "chained comparisons are not supported inside assertions" in Catch2.
+        const auto replica_number_is_1_or_2 = (1 == replica_number || 2 == replica_number);
```
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:405]: The limitation is documented here: https://github.com/catchorg/Catch2/blob/devel/docs/assertions.md#other-limitations

The compiler error message from Catch2 and option 1 in the documentation there suggest surrounding the expression in parentheses, but that did not work for me. That is, this does not work:
```c++
REQUIRE(1 == replica_number || 2 == replica_number);
```

...and this did not make things any better (at least when I tried it):
```c++
REQUIRE((1 == replica_number || 2 == replica_number));
```

Also, the author of the documentation linked above dismisses the use of `||` in assertions as a bad test ("There is no simple rewrite rule for `||`, but I generally believe that if you have `||` in your test expression, you should rethink your tests."), but I cannot think of another way to make any specific assertions about this behavior. Open to suggestions here.

File: unit_tests/src/test_replica_truncate.cpp [hunk 4]
```diff
@@ -470,6 +623,9 @@ TEST_CASE("truncate_locked_data_object__issue_7104")
     char* output_str{};
     const auto free_output_str = irods::at_scope_exit{[&output_str] { std::free(output_str); }};
 
+    constexpr char* error_message_template =
```
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:626]: Will update

File: unit_tests/src/test_replica_truncate.cpp [hunk 5]
```diff
@@ -716,47 +970,158 @@ TEST_CASE("inputs_that_will_not_work")
...
+                           {"replica_number", nlohmann::json::value_t::null},
+                           {"resource_hierarchy", nlohmann::json::value_t::null}};
+
+        CHECK(expected_json_output == json_out);
     }
 
     SECTION("DEST_RESC_HIER_STR_KW_not_allowed")
     {
-        addKeyVal(&truncate_doi.condInput, DEST_RESC_HIER_STR_KW, default_resc.c_str());
+        constexpr char* keyword = DEST_RESC_HIER_STR_KW;
```
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:992]: Will update

File: unit_tests/src/test_replica_truncate.cpp [hunk 6]
```diff
@@ -716,47 +970,158 @@ TEST_CASE("inputs_that_will_not_work")
 
     SECTION("DEST_RESC_NAME_KW_not_allowed")
     {
-        addKeyVal(&truncate_doi.condInput, DEST_RESC_NAME_KW, default_resc.c_str());
+        constexpr char* keyword = DEST_RESC_NAME_KW;
```
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:973]: Will update

File: unit_tests/src/test_replica_truncate.cpp (hunk 1 above)
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:1075]: Great, that's good to hear. Thanks!

File: unit_tests/src/test_replica_truncate.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:132]: Oh, are you saying that the `REQUIRE_NOTHROW` may not be necessary?

File: unit_tests/src/test_replica_truncate.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp]: Done

File: unit_tests/src/test_replica_truncate.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp]: Done

File: unit_tests/src/test_replica_truncate.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp]: Done

File: unit_tests/src/test_replica_truncate.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply unit_tests/src/test_replica_truncate.cpp:132]: Well, the Catch2 (v2.13.8) [documentation](https://github.com/catchorg/Catch2/blob/v2.13.8/docs/assertions.md#exceptions) for `REQUIRE_NOTHROW` mentions statements and expressions. It's not clear to me if what you have is considered proper usage. I completely understand the intent behind that line. I'm concerned whether we've _(me included)_ been misusing the macro.

For example ...
```cpp
REQUIRE_NOTHROW(x = fn_that_throws(y, z));
```
vs ...
```cpp
REQUIRE_NOTHROW((x = fn_that_throws(y, z)));
```
vs ...
```cpp
REQUIRE_NOTHROW([&] { x = fn_that_throws(y, z); }());
```
Catch2's docs explicitly show the third option as being correct.

One way to test this is by replacing `output_str` with garbage and seeing if the macro handles it.

File: unit_tests/src/test_replica_truncate.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply unit_tests/src/test_replica_truncate.cpp:405]: What you have is fine to me. I just wasn't aware of that limitation. Now I have to go read about it so I don't get confused/stuck later on.

[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:405]: Okay. It's possible that the parentheses solves the issue and I just did it wrong or something, but it didn't work for me. So if we are okay with the separate variable, then I will just resolve this.

File: unit_tests/src/test_replica_truncate.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:132]: Ooooh, I see. So I should re-do this as the third option?

**Comment by alanking** [comment]
Saw one failure in the core tests. I ran the failing test again and it passed, so I created a separate issue to look into this test: https://github.com/irods/irods/issues/7923 The test is not related to truncate at all because it has to do with `iadmin modresc rebalance`.

File: unit_tests/src/test_replica_truncate.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply unit_tests/src/test_replica_truncate.cpp:405]: No need to change what you have.

Resolving.

File: unit_tests/src/test_replica_truncate.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply unit_tests/src/test_replica_truncate.cpp:132]: You can certainly do that.

Or you tweak that line to throw and see if the `REQUIRE_NOTHROW` catches it. If it does, then what you have is fine.

[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:132]: Given that the documentation explicitly calls this out and I have confirmed that it works, let's just use the lambda expression.

I'm going to run the tweak to make this throw and see what happens before resolving this.

[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp]: I modified this very line to this:
```c++
REQUIRE_NOTHROW( json_out = nlohmann::json::parse("{") );
```

And got the following in my test report:
```xml
    <testcase classname="irods_replica_truncate.global" name="basic_two_standalone_resources/same size" time="0.026" status="run">
      <error message="json_out = nlohmann::json::parse(&quot;{&quot;)" type="REQUIRE_NOTHROW">
FAILED:
  REQUIRE_NOTHROW( json_out = nlohmann::json::parse("{") )
[json.exception.parse_error.101] parse error at line 1, column 2: syntax error while parsing object key - unexpected end of input; expected string literal
at /irods_source/unit_tests/src/test_replica_truncate.cpp:132
      </error>
    </testcase>
```

File: unit_tests/src/test_replica_truncate.cpp (hunk 1 above)
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp:1075]: Resolving as no objections have been raised to this as the expected behavior.

File: unit_tests/src/test_replica_truncate.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/src/test_replica_truncate.cpp]: I will also add that the rest of the tests ran successfully afterwards, so I guess it caught the exception? Still, I think following the recommendation in the documentation is better.

[the remark this answers is not in the document]
Comment (korydraughn) [reply unit_tests/src/test_replica_truncate.cpp]: I agree. It's simply easier to just do what the documentation recommends. Can't go wrong with that.

But it's good to know the macro **appears** to do what you intended.

**Comment by korydraughn** [comment]
Looks like this is ready?

**Comment by alanking** [comment]
Yep, I think so. All I modified since running the core tests were the unit tests and the unit tests continue to pass. Squashed.

**Review by korydraughn — APPROVED** [review]
Pound it.

**Comment by alanking** [comment]
#'d. Mergin

**Pull request merged**